app/server_inactivity.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+	"github.com/mattermost/mattermost-server/v6/store"
+)
+
+const serverInactivityHours = 100
+
+func (s *Server) doInactivityCheck() {
+	inactivityDurationHourseEnv := os.Getenv("MM_INACTIVITY_DURATION")
+	inactivityDurationHours, parseError := strconv.ParseFloat(inactivityDurationHourseEnv, 64)
+	if parseError != nil {
+		// default to 100 hours
+		inactivityDurationHours = serverInactivityHours
+	}
+
+	systemValue, sysValErr := s.Store.System().GetByName("INACTIVITY")
+	if sysValErr != nil {
+		if _, ok := sysValErr.(*store.ErrNotFound); !ok {
+			mlog.Error("An error occurred while getting INACTIVITY from system store", mlog.Err(sysValErr))
+		}
+	}
+
+	post, _ := s.Store.Post().GetLastPostRow()
+	session, _ := s.Store.Session().GetLastSessionRow()
+
+	// If we have a system value, it means this job already ran atleast once.
+	// we then check the last time the job ran plus the last time a post was made to determine if we
+	// can remind the user to use workspace again. If no post was made, we check the last time they logged in (session)
+	// and determine whether to send them a reminder.
+	if systemValue != nil {
+		sysT, _ := strconv.ParseInt(systemValue.Value, 10, 64)
+		tt := time.Unix(sysT/1000, 0)
+		timeLastSentInativityEmail := time.Since(tt).Hours()
+
+		if post != nil {
+			lastPostAt := post.CreateAt
+			posT := time.Unix(lastPostAt/1000, 0)
+			timeForLastPost := time.Since(posT).Hours()
+
+			if timeLastSentInativityEmail > inactivityDurationHours && timeForLastPost > inactivityDurationHours {
+				s.takeInactivityAction()
+			}
+			return
+		}
+
+		if session != nil {
+			lastSessionAt := session.CreateAt
+			sesT := time.Unix(lastSessionAt/1000, 0)
+			timeForLastSession := time.Since(sesT).Hours()
+
+			if timeLastSentInativityEmail > inactivityDurationHours && timeForLastSession > inactivityDurationHours {
+				s.takeInactivityAction()
+			}
+			return
+		}
+	}
```

Comment:
Wondering what is the value of checking both? If there is post activity for sure there had been session inactivity, and if there wasn't post inactivity it never checks for session inactivity. Would checking both be redundant here?

Reply:
From both the post and session, we're interested in when the *time* they last happened. Also, since MM is a chat app, checking whether people are sending posts or not should be the first indicator of inactivity/activity. But it's possible that no posts were ever made and therefore we wouldn't have a *time* reference to determine inactivity. In such a case, we use session as a fallback.